### PR TITLE
Feature/app store

### PR DESCRIPTION
**Pull request by anil-intelops**

### CODE REVIEW

File: server/pkg/store/astra/db.go [hunk 1]
```diff
@@ -25,6 +26,7 @@ func NewStore() (*AstraServerStore, error) {
 		return nil, fmt.Errorf("failed to connect to astra db, %w", err)
 	}
 	a.keyspace = a.c.Keyspace()
+	a.keyspace = "capten_server"
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:29]: no hard coding keyspace

File: server/pkg/store/astra/db.go [hunk 2]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
...
+	if err != nil {
+		return fmt.Errorf("failed to store app config : %w", err)
+	}
+
+	if appExists {
+		return fmt.Errorf("app is already available")
+	}
+
+	insertQuery := &pb.Query{
+		Cql: fmt.Sprintf("INSERT INTO %s.app_config (name, chart_name, repo_name,release_name, repo_url, namespace, version, create_namespace, privileged_namespace, launch_ui_url, launch_ui_redirect_url, category, icon, description, launch_ui_values, override_values, created_time, id) VALUES ('%s', '%s', '%s', '%s', '%s', '%s', '%s', %t, %t, '%s', '%s', '%s', '%s', '%s', '%s', '%s','%v', '%s' );",
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:344]: define as global const string above in this file

File: server/pkg/store/astra/db.go [hunk 3]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
...
+			a.keyspace, config.AppName, config.ChartName, config.RepoName, config.ReleaseName, config.RepoURL, config.Namespace, config.Version, config.CreateNamespace, config.PrivilegedNamespace, config.LaunchURL, config.LaunchRedirectURL, config.Category, config.Icon, config.Description, config.LaunchUIValues, config.OverrideValues, time.Now().Format("2006-01-02 15:04:05"), uuid.New().String()),
+	}
+
+	_, err = a.c.Session().ExecuteQuery(insertQuery)
+	if err != nil {
+		return fmt.Errorf("failed to initialise db: %w", err)
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:350]: "insert query failed"

File: server/pkg/store/astra/db.go [hunk 4]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
...
+func (a *AstraServerStore) AddAppToStore(config *types.StoreAppConfig) error {
+	appExists, err := a.isAppExistsInStore(config.AppName, config.Version)
+	if err != nil {
+		return fmt.Errorf("failed to store app config : %w", err)
+	}
+
+	if appExists {
+		return fmt.Errorf("app is already available")
+	}
+
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:342]: this check not needed.
insert will fail with app exist would be enough.

For synchronize app case 
addorupdateTostore needed, that would do upsert query like below
INSERT INTO your_table (id, column1, column2)
        VALUES (?, ?, ?)
        IF NOT EXISTS

File: server/pkg/store/astra/db.go [hunk 5]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
...
+	_, err = a.c.Session().ExecuteQuery(insertQuery)
+	if err != nil {
+		return fmt.Errorf("failed to initialise db: %w", err)
+	}
+
+	return nil
+}
+
+func (a *AstraServerStore) UpdateAppInStore(config *types.StoreAppConfig) error {
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:356]: separate update is not needed 
use common method AddOrUpdateApp

File: server/pkg/store/astra/db.go [hunk 6]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
 
 	return "", fmt.Errorf("cluster not found")
 }
+
+func (a *AstraServerStore) isAppExistsInStore(name, version string) (bool, error) {
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:314]: not needed

File: server/pkg/store/astra/db.go [hunk 7]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
...
+	fmt.Println(updateQuery)
+	_, err := a.c.Session().ExecuteQuery(updateQuery)
+	if err != nil {
+		return fmt.Errorf("failed to initialise db: %w", err)
+	}
+
+	return nil
+}
+
+func (a *AstraServerStore) DeleteAppFromStore(name, version string) error {
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:372]: remove "FromStore" "InStore" with methods, as this methods part of store

File: server/pkg/store/astra/db.go [hunk 8]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
...
+
+	deleteQuery := &pb.Query{
+		Cql: fmt.Sprintf(
+			"DELETE FROM %s.app_config WHERE name='%s' AND version='%s';",
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:376]: move to global const all queries

File: server/pkg/store/astra/db.go [hunk 9]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
...
+			a.keyspace, name, version),
+	}
+
+	_, err := a.c.Session().ExecuteQuery(deleteQuery)
+	if err != nil {
+		return fmt.Errorf("failed to initialise db: %w", err)
+	}
+
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:384]: update error message in all places

File: server/pkg/store/astra/db.go [hunk 10]
```diff
@@ -308,3 +310,291 @@ func (a *AstraServerStore) getClusterID(orgID, clusterName string) (string, erro
...
+			return nil, fmt.Errorf("failed to get repo url: %w", err)
+		}
+		cqlNamespace, err := client.ToString(row.Values[4])
+		if err != nil {
+			return nil, fmt.Errorf("failed to get Namespace: %w", err)
+		}
+		cqlVersion, err := client.ToString(row.Values[5])
+		if err != nil {
+			return nil, fmt.Errorf("failed to get version: %w", err)
+		}
```
Comment (vramk23) [inline server/pkg/store/astra/db.go:536]: reading DB values can be optimized for both gep app methods with common method convert resultset to struct

File: server/pkg/store/cassandra/db.go [hunk 11]
```diff
@@ -188,3 +194,66 @@ func (c *CassandraServerStore) DeleteCluster(orgID, clusterName string) error {
...
+
+	var config types.AppConfig
+	iter.Scan(&config)
+	if config.Name != "" {
+		return false
+	}
+	return true
+}
+
+func (c *CassandraServerStore) AddAppToStore(config *types.StoreAppConfig) error {
```
Comment (vramk23) [inline server/pkg/store/cassandra/db.go:211]: same comments applicable in astra, check for cassandra as well

File: server/pkg/api/store_apps.go [hunk 12]
```diff
@@ -4,29 +4,224 @@ import (
...
-	return &serverpb.AddStoreAppResponse{}, nil
+
+	if request.AppConfig.AppName == "" {
+		s.log.Errorf("failed to add app config to store, %v", "App name is missing")
+		return &serverpb.AddStoreAppResponse{
+			Status:        serverpb.StatusCode_INTERNRAL_ERROR,
+			StatusMessage: "failed add app config to store, app name is missing",
+		}, nil
+	} else if request.AppConfig.Version == "" {
+		s.log.Errorf("failed to add app config to store, %v", "App version is missing")
```
Comment (vramk23) [inline server/pkg/api/store_apps.go:20]: check with one condition both appname and version are not nil.
send error message "appName or version is empty"
do the same for other methods

File: proto/server.proto [hunk 13]
```diff
@@ -206,12 +206,12 @@ message StoreAppConfig {
   string namespace = 9;
   bool createNamespace = 10;
   bool privilegedNamespace = 11;
-  bytes icon = 12;
+  string icon = 12;
   string launchURL = 13;
```
Comment (vramk23) [inline proto/server.proto:210]: this should be byte array or BLOB

**Review by vramk23 — CHANGES_REQUESTED** [review]

File: server/pkg/store/astra/db.go (hunk 1 above)
Comment (anil-intelops) [reply server/pkg/store/astra/db.go:29]: removed

**Pull request merged**